### PR TITLE
fix: ensure buffers are retained when copying an already buffered table

### PR DESCRIPTION
**Pull request by onelson**
Refs #4780

When the tables sent into `tableFind` include a yield, the result generated by the yield can cause the tables to be released before `tableFind` can use them.

`tableFind` will make its own copy, but prior to this change the copy operation would hand back the input directly when it was already buffered. In effect, this means we produce references to the same underlying buffers without us incrementing the ref count.

~~With this diff, for the case where we try to copy an already buffered table, we call its own `Copy()` method which will `Retain()` each of the internal buffers.~~

After a few twists and turns, the fixes for this issue came from:

- making `tableBuffer`'s `Do` and `Done` idempotent such that calling `Done` after `Do` would not release the underlying arrow arrays repeatedly.
- ensuring `table.Copy()` always calls `Retain()` on the underlying arrow arrays (_previously this was not the case when the table is already buffered_).

A `tableBuffer` could be spent by calling either `Do()` or `Done()`, then later we might call `Do()` again. Prior to this patch, `Do()` would silently do nothing. Instead, it will now return an error "table already read." This is in line with other implementations of the Table APIs.

Similarly, `Copy()` will now panic if called on a table that is already "spent." Prior to the explicit panic, we'd see panics surfaced by invalid (`nil` pointer) access to underlying arrow arrays somewhere deeper into the execution of the program. _Seems as though this is a case where a `tableBuffer` lived too long via programmer error so a more immediate/explicit panic makes sense._

---

N.b this is not a total fix for #4780. We still have a separate problem related to how `tableFind` will use a random result from the input table stream when there is >1 result to be had. #4813 should address the remaining problems surfaced by the initial report.

### Done checklist
- [x] docs/SPEC.md updated **N/A**
- [x] Test cases written

### CODE REVIEW

**Comment by onelson** [comment]
I had hoped to add a flux acceptance test matching the repro used while investigating #4780 but this is not possible to do since any non-empty result in a test case will mean the test will fail. Once #4300 has been completed, this should be possible.

File: execute/table/copy.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ import "github.com/influxdata/flux"
 // this happens and prevent it.
 func Copy(t flux.Table) (flux.BufferedTable, error) {
 	if tbl, ok := t.(flux.BufferedTable); ok {
-		return tbl, nil
+		return tbl.Copy(), nil
```
Comment (onelson) [inline execute/table/copy.go:19]: This might be a question for @jsternberg.

To mitigate the issue, we can either call `Copy()` as in the diff currently, or we can build a fresh `BufferedTable` by removing this block entirely. The semantic difference that I can see is the former appears to carry the value for `BufferedTable.i` forward, whereas the later would reset it to `0`.

I was not clear on the impact either way, but since `BufferedTable.Copy` is already a thing, I hoped that it accounted for this already.

Comment (jsternberg) [reply execute/table/copy.go:19]: These functions are weird and the naming to them is a bit of a mistake. `execute.CopyTable()` can actually be thought of more as "ensure table is buffered" while `buf.Copy()` is "copy this table so it can be used somewhere else".

The difference is that `execute.CopyTable()` _consumes_ its input. So the test in there so see "is this already a buffered table?" is mostly meant as an optimization to the function's true purpose. `buf.Copy()` will produce a copy of the table so it can be used in multiple places.

The naming isn't the best but this change would cause `execute.CopyTable()` to not work properly in the locations where it is presently used.

Comment (onelson) [reply execute/table/copy.go:19]: Based on your description here, perhaps the thing to do is make the `Copy` call closer (hopefully inside) `tableFind`. There's a slight bit of indirection via a function named `objects.NewTable`.

Regardless, I can look for another point in which to make this happen if this is contradictory to the design.

Thoughts?

Comment (onelson) [reply execute/table/copy.go:19]: I can add a type assertion to `tableFind` such that when it gets an already buffered table it calls `Copy()` on it before we hand it off to `NewTable()`. This seems sufficient to fix the bug for the identified case, but it'll put me back to square one as far as testing goes :grimacing:

**Comment by jsternberg** [comment]
I had a look with this with @onelson and we discovered something I wasn't really aware of. It appears like the table find code is invoking everything correctly as-is (other than the multiple result thing). The problem seems to be that the `tableBuffer` struct, which is used for buffering data like is done here, has a bug with its `Done()` method.

`Done()` is meant to be idempotent and you can call multiple times. The table find code uses it that way. Unfortunately, there appears to be a bug in `tableBuffer` where if you invoke `Do()` (such as to read the table buffer) and then you call `Done()`, it decrements the reference count twice instead of only once as it is supposed to do.

All of the other implementations of `flux.Table` follow this pattern but `tableBuffer` isn't. I believe that table find is using the Go API in the proper way and `tableBuffer` is just not fulfilling the API the way it is supposed to be.

File: execute/table/copy.go (hunk 1 above)
Comment (onelson) [reply execute/table/copy.go:19]: I just paired with Jonathan briefly and with his guidance I'm looking at an alternate fix.

Sounds like the implementations for `Do()` and `Done()` are not currently idempotent and they need to be, and it seems like if we truly do need to drain the table in question we need to add a missing `Copy()` (but there is doubt that this is actually needed today).

**Comment by onelson** [comment]
Moving this back into draft while I rework the patch.

**Comment by onelson** [comment]
Still have more digging to do. After adding atomic guards around `Do`/`Done` for `tableBuffer` I see a different panic here:

https://github.com/influxdata/flux/blob/master/stdlib/universe/table_fns.go#L261

Looks like the length check panics if the underlying arrow array has already been released, so either I screwed up the guard or there's something else contributing. Eyeballing the field I am using for the guard, `used` is 0 and 1 when I'd expect it to be, so whatever is happening is unclear.

I'll have to pick this back up next week.

File: execute/table/copy.go (hunk 1 above)
Comment (onelson) [reply execute/table/copy.go]: > To mitigate the issue, we can either call Copy() as in the diff currently, or we can build a fresh BufferedTable by removing this block entirely.

After some back and forth, and the discovery of those other idempotent-related issues, the conclusion was to remove the block rather than artificially add a `Copy()` call.

What I didn't realize when I posed the question in the first place is there's a call here which will copy the table already:

https://github.com/influxdata/flux/blob/878a9f1a60837db0e28c473ea0e12a1f730650ed/stdlib/universe/table_fns.go#L251

Adding an extra copy would have been redundant, possibly leading to a leak (though this wasn't clear from the test suite).

File: execute/table/copy.go [hunk 2]
```diff
@@ -89,13 +101,16 @@ func (tb *tableBuffer) BufferN() int {
 }
 
 func (tb *tableBuffer) Copy() flux.BufferedTable {
-	for i := tb.i; i < len(tb.buffers); i++ {
+	if tb.used == 1 {
```
Comment (jsternberg) [inline execute/table/copy.go:104]: Use the atomic load here.

Comment (onelson) [reply execute/table/copy.go]: Whoops! Makes sense.

File: execute/table/copy.go [hunk 3]
```diff
@@ -89,13 +101,16 @@ func (tb *tableBuffer) BufferN() int {
 }
 
 func (tb *tableBuffer) Copy() flux.BufferedTable {
-	for i := tb.i; i < len(tb.buffers); i++ {
+	if atomic.LoadInt32(&tb.used) == 1 {
+		panic("tried to copy an already used tableBuffer")
```
Comment (jsternberg) [inline execute/table/copy.go:105]: I'm going to approve, but is there any chance you can have this panic only exist when the `assert` tag is used?

The `assert` tag is part of arrow for multiple releases and we use it when running tests.

**Review by jsternberg — APPROVED** [review]

File: execute/table/copy.go (hunk 3 above)
Comment (onelson) [reply execute/table/copy.go:105]: I'll look into that. You'd rather see some later downstream error or panic in these cases, assuming they reach production somehow?

Comment (onelson) [reply execute/table/copy.go]: I switched this to be a logging call instead for now. Arrow's `debug.Assert()` is neat, but it lives in their `internal/` module space so I don't think we can access it.

We might want to add something like that to flux directly.

File: execute/table/copy.go [hunk 4]
```diff
@@ -89,13 +102,20 @@ func (tb *tableBuffer) BufferN() int {
 }
 
 func (tb *tableBuffer) Copy() flux.BufferedTable {
-	for i := tb.i; i < len(tb.buffers); i++ {
+
+	// TODO: make this panic instead of just logging when running tests.
+	//  Arrow has a `debug.Assert` thing which will noop unless the assert build
+	//  tag is set. We could do something similar.
+	if atomic.LoadInt32(&tb.used) == 1 {
+		log.Println("tried to copy an already used tableBuffer")
```
[the remark this answers is not in the document]
Comment (onelson) [reply execute/table/copy.go:110]: I saw a random usage of `log.Println()` over in `compiler/runtime.go` :shrug:

I just didn't want to pile on adding something like that with this PR. Would rather bikeshed it separately if needed.

[the remark this answers is not in the document]
Comment (onelson) [reply execute/table/copy.go]: I was hesitant to add this immediately since verification seemed tricky. Added with b8c5a74786e86e93a54da5e2703323023db94eb1

I weakly tested by flipping the assertion condition and adding/removing the build tag in the Makefile to make sure panics happened or didn't happen when we wanted.

File: Makefile [hunk 5]
```diff
@@ -12,7 +12,7 @@ export PKG_CONFIG:=$(shell pwd)/pkg-config.sh
 
 export GOOS=$(shell go env GOOS)
 export GO_BUILD=env GO111MODULE=on go build $(GO_ARGS)
-export GO_TEST=env GO111MODULE=on go test $(GO_ARGS)
```
Comment (jsternberg) [inline Makefile:15]: I believe this is defined in `.circleci/config.yml`. We inject `assert` through an argument.

**Review by jsternberg — APPROVED** [review]
One minor revert to the Makefile but otherwise good.

File: Makefile (hunk 5 above)
Comment (onelson) [reply Makefile:15]: Seems like if we go that route, we won't get the nice feedback when testing locally. I'm fine with reverting it, but it seems a bit odd :woozy_face:
